src/proof_rules/calculus/soundness_checker.rs

```diff
@@ -1,121 +1,310 @@
+impl IntoIterator for ApproximationList {
+    type Item = ApproximationRecord;
+    type IntoIter = std::vec::IntoIter<ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
 
-#[derive(Debug, Clone, Default)]
-pub struct SoundnessBlame {
-    pub soundness: Soundness,
-    pub approximations: ApproximationList,
-    pub calculus: Option<Calculus>,
+impl<'a> IntoIterator for &'a ApproximationList {
+    type Item = &'a ApproximationRecord;
+    type IntoIter = std::slice::Iter<'a, ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
 }
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
-pub enum ApproximationKind {
-    #[default]
-    Exact,
-    Under,
-    Over,
-    Unknown,
+impl<'a> IntoIterator for &'a mut ApproximationList {
+    type Item = &'a mut ApproximationRecord;
+    type IntoIter = std::slice::IterMut<'a, ApproximationRecord>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter_mut()
+    }
 }
 
-impl ApproximationKind {
-    pub fn supremum(self, other: ApproximationKind) -> ApproximationKind {
-        match self.partial_cmp(&other) {
-            Some(std::cmp::Ordering::Equal) => self,
-            Some(std::cmp::Ordering::Less) => other,
-            Some(std::cmp::Ordering::Greater) => self,
-            None => ApproximationKind::Exact, // If they are incomparable, return Exact (which is the top element)
+impl ApproximationList {
+    pub fn infimum(&self) -> ApproximationKind {
+        self.0
+            .iter()
+            .fold(ApproximationKind::EXACT, |acc, approx_entry| {
+                acc & approx_entry.kind
+            })
+    }
+
+    // convenience helpers
+    pub fn push(&mut self, rec: ApproximationRecord) {
+        self.0.push(rec)
+    }
+
+    pub fn iter(&self) -> std::slice::Iter<'_, ApproximationRecord> {
+        self.0.iter()
+    }
+}
+
+impl From<Vec<ApproximationRecord>> for ApproximationList {
+    fn from(v: Vec<ApproximationRecord>) -> Self {
+        ApproximationList(v)
+    }
+}
+
+impl std::ops::Deref for ApproximationList {
+    type Target = Vec<ApproximationRecord>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for ApproximationList {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct ProcSoundness {
+    sound_proofs: bool,
+    sound_refutations: bool,
+    approximations: ApproximationList,
+    calculus: Option<Calculus>,
+}
+
+impl ProcSoundness {
+    /// Create a new ProcSoundness instance by analyzing the given approximations and calculus in the given direction.
+    pub fn new(
+        direction: Direction,
+        approximations: ApproximationList,
+        calculus: Option<Calculus>,
+    ) -> Self {
+        let approx = approximations.infimum();
+        // The mapping between approximation kinds and soundness is as follows:
```

Comment:
It *cannot* be this complicated. In particular, the XORs are impossible to read and explain. I suggest explicit if-then-else chains that make the code overall readable.